feat: Add `convertCurrency` HogQL function

rafaeelaudibert opened the pull request:
This function is meant to be used to convert from a specific currency to another for a given date. This uses the newly-created `exchange_rate_dict` to trigger the conversion. We don't have dates for every single day yet, but they're coming soon, this should be enough to test it out in prod!

Also did it some extra stuff and actually exposed the `exchange_rate` table to the world! This might help customers do whatever they want with the data.

## Does this work well for both Cloud and self-hosted?

Yuup

## How did you test this code?

Manually + unit tests

rafaeelaudibert committed:
feat: Add `hogql_convertCurrency` HogQL function

This function is meant to be used to convert from a specific currency to
another for a given date. This uses the newly-created
`exchange_rate_dict` to trigger the conversion. We don't have dates for
every single day yet, but they're coming soon, this should be enough to
test it out in prod!

rafaeelaudibert commented on posthog/hogql/printer.py:
```diff
@@ -1208,6 +1209,12 @@ def visit_call(self, node: ast.Call):
                     return f"dictGetOrNull('channel_definition_dict', 'type_if_organic', (coalesce({args[0]}, ''), 'medium'))"
+                elif (
+                    node.name == "hogql_convertCurrency"
+                ):  # hogql_convertCurrency(from_currency, to_currency, amount, timestamp)
+                    from_currency, to_currency, amount, *_rest = args
+                    date = f"toDate({args[3]})" if len(args) > 3 and args[3] else "today()"
+                    return f"multiplyDecimal(divideDecimal({amount}, dictGet({EXCHANGE_RATE_DICTIONARY_NAME}, 'rate', {from_currency}, {date}, 0)), dictGet({EXCHANGE_RATE_DICTIONARY_NAME}, 'rate', {to_currency}, {date}, 0))"
```
It's checked above

rafaeelaudibert commented on posthog/hogql/printer.py:
```diff
@@ -1208,6 +1209,12 @@ def visit_call(self, node: ast.Call):
+                    node.name == "hogql_convertCurrency"
+                ):  # hogql_convertCurrency(from_currency, to_currency, amount, timestamp)
+                    from_currency, to_currency, amount, *_rest = args
+                    date = f"toDate({args[3]})" if len(args) > 3 and args[3] else "today()"
```
There is, that's what the last argument returns, we'll return 0. I've checked that `div(something, 0)` is a valid operation in Clickhouse

rafaeelaudibert commented on posthog/hogql/functions/mapping.py:
```diff
@@ -1156,11 +1157,38 @@ def compare_types(arg_types: list[ConstantType], sig_arg_types: tuple[ConstantTy
+                DecimalType(),
+            ),
+            (
+                (
+                    StringType(),
+                    StringType(),
+                    DecimalType(),
+                    StringType(),
+                ),
+                DecimalType(),
+            ),
```
I wanna accept a string, but it's correct I could accept a `DateTime` (or even a date) here instead

rafaeelaudibert committed:
fix: Accept actual dates rather than string

This will make our lives much easier

rafaeelaudibert committed:
fix: Remove `data_type` because I can't type it right

rafaeelaudibert committed:
fix: Make it unknown god

rafaeelaudibert committed:
fix: Add `DecimalType` to `AnyConstantType`

lricoy reviewed (approved):
HogQL is now one step ahead of google finance in Brazil: https://rodolphoarruda.pro.br/google-finance-in-brazil/

Approving as it LGTM and to unblock, but as always, wait for a more experienced reviewer if you think it is appropriate

robbie-c commented:
> HogQL is now one step ahead of google finance in Brazil: [rodolphoarruda.pro.br/google-finance-in-brazil](https://rodolphoarruda.pro.br/google-finance-in-brazil/)

Huh, that's fascinating. We're likely to be asked what the source of our data is, we should include this in our docs

rafaeelaudibert commented:
@mariusandra As the resident HogQL expert mind giving this a look? It's a rather small PR

robbie-c commented on posthog/hogql/test/test_printer.py:
```diff
@@ -2031,6 +2032,38 @@ def test_lookup_organic_medium_type(self):
+            printed,
+        )
+
+    def test_currency_conversion_without_date(self):
```
Can we include some tests where the currency property is null / undefined / garbage?

It's fine if we don't do anything sensible if there's garbage, but would be good to at least have the test.

robbie-c replied:
Oh I'm being dumb, we're not actualy checking the result

robbie-c replied:
Can you add some tests to test_query where this function is called and we check the result?

rafaeelaudibert replied:
Yeah, sure, let me add that 🚧

I think the printer will just print it, though? We'd benefit from an actual test, but where would that go? I assume it might be a better fit once we're actually using this on Web Analytics

rafaeelaudibert replied:
Ah, I didn't know about `test_query.py`, let me see what I can do there

rafaeelaudibert replied:
It's there now, and I have one of the coolest tests ever :)

rafaeelaudibert commented:
> I can leave a proper review on Monday (OOO today/tomorrow), though my main question is: why the "hogql_" in front? We are calling it PostHog SQL to external users now as well... So is just drop that, unless CH has a function with the same name. If not, all good: we can introduce and have introduced our own functions

Copied from the channelType functions, but happy to remove that :)

mariusandra commented:
I think we only use those internally in our own expressions, and users don't write those strings themselves? I could be wrong though..

rafaeelaudibert committed:
feat: Introduce `test_query` tests

I had to poke around and change a lot of stuff to get this to work, but it's working now, nice

rafaeelaudibert committed:
refactor: Rename function to `convertCurrency`

No need to use HogQL in front of it

rafaeelaudibert commented on posthog/hogql/functions/mapping.py:
```diff
@@ -399,6 +401,7 @@ def compare_types(arg_types: list[ConstantType], sig_arg_types: tuple[ConstantTy
     "_toUInt128": HogQLFunctionMeta("toUInt128", 1, 1),
     "toFloat": HogQLFunctionMeta("accurateCastOrNull", 1, 1, suffix_args=[ast.Constant(value="Float64")]),
     "toDecimal": HogQLFunctionMeta("toDecimal64OrNull", 2, 2),
+    "DATE": HogQLFunctionMeta("toDate", 1, 1),
```
I need something that can convert from a string to a `Date` rather than a `Nullable(Date)`. Clickhouse has a `DATE` function so I thought it made sense to call it this, but I'm open to suggestions

robbie-c replied:
There's a big project that @oliverb123 is working on to add a type system, so this shouldn't be needed in the future.

For now, I would probably call this something like `_toDateNotNull`

rafaeelaudibert commented on posthog/hogql/database/database.py:
```diff
@@ -153,6 +155,7 @@ class Database(BaseModel):
         "sessions",
         "heatmaps",
         "query_log",
+        "exchange_rate",
```
You can query the `exchange_rate` table now! I think this will be a fun addition, no reason not to do it

rafaeelaudibert commented on posthog/hogql/printer.py:
```diff
@@ -491,6 +494,7 @@ def visit_join_expr(self, node: ast.JoinExpr) -> JoinExprResponse:
                 self.dialect == "clickhouse"
                 and not isinstance(table_type.table, FunctionCallTable)
                 and not isinstance(table_type.table, SavedQuery)
+                and not isinstance(table_type.table, ExchangeRateTable)
```
Gotta skip adding the `team_id` clause to queries against this table or else it breaks catastrophically

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

robbie-c commented on posthog/hogql/test/test_query.py:
```diff
@@ -1581,3 +1584,65 @@ def test_db_created_once(self):
+        response = execute_hogql_query(query, team=self.team)
+        self.assertEqual(response.results, [(Decimal("100.00"),)])
+
+    def test_currency_conversion_super_nested(self):
```
this is awesome!

rafaeelaudibert commented:
@robbie-c How do you feel about merging this now and letting me have something testable in prod right now? I'm confident on this and marius can give us a review on Monday and I can fix it forward
